lib/CFRPageActions.jsm

```diff
@@ -0,0 +1,229 @@
+
+    this.label.value = notificationText;
+
+    // We can safely get the width of the label even though the recommendation is
+    // collapsed; the label itself remains full width (with its overflow hidden)
+    const [{width}] = this.label.getClientRects();
+    this.urlbar.style.setProperty("--cfr-label-width", `${width}px`);
+
+    if (shouldExpand) {
+      this._clearScheduledStateChanges();
+
+      // After one second, expand
+      this._expand(DELAY_BEFORE_EXPAND_MS);
+
+      // Five seconds later, collapse again
+      this._collapse(DELAY_BEFORE_EXPAND_MS + DURATION_OF_EXPAND_MS);
+    }
+  }
+
+  hide() {
+    this.container.hidden = true;
+    this._collapse();
+  }
+
+  _expand(delay = 0) {
+    if (!delay) {
+      // Non-delayed state change overrides any scheduled state changes
+      this._clearScheduledStateChanges();
+      this.urlbar.setAttribute("cfr-recommendation-state", "expanded");
+    } else {
+      this.stateTransitionTimeoutIDs.push(this.window.setTimeout(() => {
+        this.urlbar.setAttribute("cfr-recommendation-state", "expanded");
+      }, delay));
+    }
+  }
+
+  _collapse(delay = 0) {
+    if (!delay) {
+      // Non-delayed state change overrides any scheduled state changes
+      this._clearScheduledStateChanges();
+      this.urlbar.setAttribute("cfr-recommendation-state", "collapsed");
+    } else {
+      this.stateTransitionTimeoutIDs.push(this.window.setTimeout(() => {
+        this.urlbar.setAttribute("cfr-recommendation-state", "collapsed");
+      }, delay));
+    }
+  }
+
+  _clearScheduledStateChanges() {
+    while (this.stateTransitionTimeoutIDs.length > 0) {
+      // clearTimeout is safe even with invalid/expired IDs
+      this.window.clearTimeout(this.stateTransitionTimeoutIDs.pop());
+    }
+  }
+
+  // This is called when the popup closes as a result of interaction _outside_
+  // the popup, e.g. by hitting <esc>
+  _popupStateChange(state) {
+    if (["dismissed", "removed"].includes(state)) {
+      this._collapse();
+    }
+  }
+
+  /**
+   * Respond to a user click on the recommendation by showing a doorhanger/
+   * popup notification
+   */
+  _handleClick(event) {
+    const browser = this.window.gBrowser.selectedBrowser;
+    if (!RecommendationMap.has(browser)) {
+      // There's no recommendation for this browser, so the user shouldn't have
+      // been able to click
+      this.hide();
+      return;
+    }
+    const {recommendation} = RecommendationMap.get(browser);
+
+    // The recommendation should remain either collapsed or expanded while the
+    // doorhanger is showing
+    this._clearScheduledStateChanges();
+
+    // A hacky way of setting the popup anchor outside the usual url bar icon box
+    // See https://searchfox.org/mozilla-central/rev/847b64cc28b74b44c379f9bff4f415b97da1c6d7/toolkit/modules/PopupNotifications.jsm#42
+    browser.cfrpopupnotificationanchor = this.container;
+
+    const {primary, secondary} = recommendation.buttons;
```

Comment:
all of these properties actually properties of `recommendation.content`, not `recommendation` itself